lib/src/license_detection/license.dart

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:meta/meta.dart';
+import 'package:pana/src/license_detection/tokenizer.dart';
+
+@sealed
+class License {
+  /// Name of the license, is empty in case of unknown license.
+  final String identifier;
+
+  /// Original text from the license file.
+  final String content;
+
+  /// Normalized [Token]s created from the original text.
+  final List<Token> tokens;
+
+  /// A map of tokens and their count.
+  final Map<String, int> occurences;
+
+  License._(this.content, this.tokens, this.occurences, this.identifier);
+
+  factory License.parse(String identifier, String content) {
+    final tokens = tokenize(content);
+    final table = generateFrequencyTable(tokens);
+    return License._(content, tokens, table, identifier);
+  }
+}
+
+/// Contains deatils regarding the results of corpus license match with unknwown text.
+@sealed
+class LicenseMatch {
+  /// Sequence of tokens that were found in the unknown
+  ///  text that matched to tokens in any of the corpus license.
+  final List<Token> tokens;
+
+  /// Confidence score of the detected license.
+  final double confidence;
+
+  /// SPDX license which matched with input.
+  final License license;
+
+  LicenseMatch(this.tokens, this.confidence, this.license);
+}
+
+/// Genearates a frequency table for the give list of tokens.
+@visibleForTesting
+Map<String, int> generateFrequencyTable(List<Token> tokens) {
+  var table = <String, int>{};
+
+  for (var token in tokens) {
+    table[token.value] = table.putIfAbsent(token.value, () => 0) + 1;
+  }
+
+  return table;
+}
+
+/// Creates [License] instances for all the corpus licenses.
+List<License> loadLicensesFromDirectories(List<String> directories) {
+  var licenses = <License>[];
+  final length = directories.length;
+
+  for (var i = 0; i < length; i++) {
+    final dir = Directory(directories[i]);
+
+    dir.listSync(recursive: false).forEach((element) {
+      final license = licenseFromFile(element.path);
+      licenses.addAll(license);
+    });
+  }
+
+  return List.unmodifiable(licenses);
+}
+
+/// Returns [License] instance for the given license file.
+@visibleForTesting
+List<License> licenseFromFile(String path) {
+  var licenses = <License>[];
+  final file = File(path);
+
+  final fileName = file.uri.toString();
+  if (!fileName.endsWith('.txt')) {
+    return <License>[];
+  }
+
+  final identifier = file.uri.pathSegments.last.split('.txt').first;
+  final content = file.readAsStringSync();
```

Comment:
So a function like `file.readAsStringSync` will do  `file.readAsBytesSync` and then decode with `utf8.decode` from `dart:convert`.

If the contents of the file is not valid UTF-8, then `utf8.decode` will throw `FormatException`. Maybe we should catch such an exception and throw an `ArgumentError`. If the license file we are given is not in UTF-8 format, then it's a programmer error, right?

We could also throw a custom exception, or even throw a new `FormatException`.. but if we want to throw `FormatException` it makes sense that we throw an exception saying which file we failed to load.

-------

Also if we throw a `FormatException` if the file is invalid, then we should document this behavior. It might make sense for `loadLicensesFromDirectories` to have a documentation comment that says:
> Throws `FormatException` if any of the directories contains a file that is not a valid license file.

Then we can throw `FormatException` if UTF-8 decoding fails, if a file is not on the form `<spdx-license-id>.txt`, or any other issue we have because the format of the file is bad.